### PR TITLE
Mutter eglstreams

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Finish https://github.com/NixOS/nixpkgs/pull/60100.

I haven't been able to test this as I lack the hardware.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: nixos/modules/services/x11/display-managers/gdm.nix [hunk 1]
```diff
@@ -88,11 +88,21 @@ in
       wayland = mkOption {
         default = true;
         description = ''
-          Allow GDM run on Wayland instead of Xserver
+          Allow GDM run on Wayland instead of Xserver.
```
Comment (worldofpeace) [inline nixos/modules/services/x11/display-managers/gdm.nix]:
```suggestion
          Allow GDM to run on Wayland instead of Xserver.
```

**Comment by worldofpeace** [comment]
@GrahamcOfBorg eval

**Comment by andersk** [comment]
Hi, thanks for packaging this! I tested with a GTX 1080 and it’s working for me. `echo $XDG_SESSION_TYPE` reports `wayland`, and `pgrep -l X` just reports an `Xwayland` process.

I haven’t figured out how to get `firefox-wayland` to use Wayland mode: even with `MOZ_ENABLE_WAYLAND=1`, `about:support` still reports “Window Protocol: x11”. But that might be due to upstream blacklists or something. I dunno.

Is there anything particular you want me to test?

**Comment by worldofpeace** [comment]
> Hi, thanks for packaging this! I tested with a GTX 1080 and it’s working for me. `echo $XDG_SESSION_TYPE` reports `wayland`, and `pgrep -l X` just reports an `Xwayland` process.
> 
> I haven’t figured out how to get `firefox-wayland` to use Wayland mode: even with `MOZ_ENABLE_WAYLAND=1`, `about:support` still reports “Window Protocol: x11”. But that might be due to upstream blacklists or something. I dunno.
> 
> Is there anything particular you want me to test?

As long as the PR was tested on these commits with this config
* https://discourse.nixos.org/t/need-gnome-3-tester-for-nvidia-eglstream-support/2904/2?u=worldofpeace

that sounds perfect. Thank you :sparkles:

**Comment by worldofpeace** [comment]
> I haven’t figured out how to get firefox-wayland to use Wayland mode: even with MOZ_ENABLE_WAYLAND=1, about:support still reports “Window Protocol: x11”. But that might be due to upstream blacklists or something. I dunno.

I don't think the firefox package in nixpkgs supports wayland yet.
We may need to look at the work Fedora has done to have this integrated well.

**Comment by andersk** [comment]
> As long as the PR was tested on these commits with this config
> * https://discourse.nixos.org/t/need-gnome-3-tester-for-nvidia-eglstream-support/2904/2?u=worldofpeace

Yes.

> I don't think the firefox package in nixpkgs supports wayland yet.

We have a `firefox-wayland` package which is what I tried, but since it’s a thin wrapper around `firefox` it’s possible that it’s still not built with the right options.

Also I’ve noticed these messages from `journalctl --user -b 0 -t gnome-shell`:

```
gnome-shell[2106]: xwayland glamor: GBM backend (default) is not available
gnome-shell[2106]: Missing Wayland requirements for glamor GBM backend
gnome-shell[2106]: glamor: 'wl_eglstream_display' not supported
gnome-shell[2106]: Missing Wayland requirements for glamor EGLStream backend
gnome-shell[2106]: Failed to initialize glamor, falling back to sw
```

and `xdpyinfo` does not show the `GLX` extension.

**Comment by marcus7070** [comment]
I just tried to build adapta-gtk-theme and somehow egl-wayland was a dependency of it. 

Anyhow, it failed to build on my machine with 

```
marcus@sevens> nix-build -A egl-wayland                                                            ~/tempbuild/nixpkgs
these derivations will be built:
  /nix/store/iq61rj6j753wx0r9v80ffp6b42rvmpy8-egl-wayland-1.1.4.drv
building '/nix/store/iq61rj6j753wx0r9v80ffp6b42rvmpy8-egl-wayland-1.1.4.drv'...
unpacking sources
unpacking source archive /nix/store/185x4vczf186q98vrxlz3c766d28ycrs-source
source root is source
patching sources
configuring
meson flags: --buildtype=plain         --libdir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/lib --libexecdir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/libexec         --bindir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/bin --sbindir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/sbin         --includedir=/nix/store/94n6a7s5pz637hv31dl22ixbssq6i2a7-egl-wayland-1.1.4-dev/include         --mandir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/share/man --infodir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/share/info         --localedir=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4/share/locale         -Dauto_features=enabled         -Dwrap_mode=nodownload         --prefix=/nix/store/5qz33mfqj9n6lmd3fbbkccy00xm458sm-egl-wayland-1.1.4
The Meson build system
Version: 0.51.2
Source dir: /build/source
Build dir: /build/source/build
Build type: native build
DEPRECATION: Duplicated values in array option is deprecated. This will become a hard error in the future.
DEPRECATION: Duplicated values in array option is deprecated. This will become a hard error in the future.
DEPRECATION: Duplicated values in array option is deprecated. This will become a hard error in the future.
Project name: wayland-eglstream
Project version: 1.1.4
C compiler for the host machine: /nix/store/l2sqva7kzr8y68lji2aajk2s8017jyq4-gcc-wrapper-8.3.0/bin/cc (gcc 8.3.0 "gcc (GCC) 8.3.0")
DEPRECATION: Duplicated values in array option is deprecated. This will become a hard error in the future.
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /nix/store/dnxp63231krq8l87kyhkbwic3w9s1pd9-pkg-config-0.29.2/bin/pkg-config (0.29.2)
Run-time dependency eglexternalplatform found: YES 1.1
Run-time dependency wayland-server found: YES 1.17.0
Run-time dependency wayland-client found: YES 1.17.0
Run-time dependency wayland-egl-backend found: YES 3
Run-time dependency threads found: YES
Found pkg-config: /nix/store/dnxp63231krq8l87kyhkbwic3w9s1pd9-pkg-config-0.29.2/bin/pkg-config (0.29.2)
Build-time dependency wayland-scanner found: YES 1.17.0
Program /nix/store/h0b47l5z1fqxcgxnh87j4dj72l71zn8h-wayland-1.17.0/bin/wayland-scanner found: YES (/nix/store/h0b47l5z1fqxcgxnh87j4dj72l71zn8h-wayland-1.17.0/bin/wayland-scanner)
Configuring wayland-eglstream.pc using configuration
Configuring wayland-eglstream-protocols.pc using configuration
Checking for function "dlsym" : NO
Library dl found: YES
Compiler for C supports arguments -Wpedantic: YES
Build targets in project: 7
Option buildtype is: plain [default: debugoptimized]
Found ninja-1.9.0 at /nix/store/sdh59gf21q30k2pyjmhq85gnajsiji5d-ninja-1.9.0/bin/ninja
meson: enabled parallel building
building
build flags: -j4 -l4
[6/18] Compiling C object 'src/25a6634@@nvidia-egl-wayland@sha/wayland-egldevice.c.o'.and-eglstream_wayland-eglstream-controller-protocol.c.o'.
FAILED: src/25a6634@@nvidia-egl-wayland@sha/wayland-egldevice.c.o
/nix/store/l2sqva7kzr8y68lji2aajk2s8017jyq4-gcc-wrapper-8.3.0/bin/cc -Isrc/25a6634@@nvidia-egl-wayland@sha -Isrc -I../src -I../include -I../wayland-egl -Iwayland-eglstream -I/nix/store/scp3wbv44ch6mdrjjhf1d3x0qnb41il6-eglexternalplatform-1.1/include -I/nix/store/h0b47l5z1fqxcgxnh87j4dj72l71zn8h-wayland-1.17.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=gnu99 -Wall -Werror -fvisibility=hidden -DWL_HIDE_DEPRECATED -Wno-pedantic -fPIC -pthread -MD -MQ 'src/25a6634@@nvidia-egl-wayland@sha/wayland-egldevice.c.o' -MF 'src/25a6634@@nvidia-egl-wayland@sha/wayland-egldevice.c.o.d' -o 'src/25a6634@@nvidia-egl-wayland@sha/wayland-egldevice.c.o' -c ../src/wayland-egldevice.c
In file included from ../src/wayland-egldevice.c:23:
../include/wayland-egldevice.h:26:10: fatal error: EGL/egl.h: No such file or directory
 #include <EGL/egl.h>
          ^~~~~~~~~~~
compilation terminated.
[9/18] Compiling C object 'src/25a6634@@nvidia-egl-wayland@sha/wayland-thread.c.o'.ayland-eglstream_wayland-eglstream-controller-protocol.c.o'.
FAILED: src/25a6634@@nvidia-egl-wayland@sha/wayland-thread.c.o
/nix/store/l2sqva7kzr8y68lji2aajk2s8017jyq4-gcc-wrapper-8.3.0/bin/cc -Isrc/25a6634@@nvidia-egl-wayland@sha -Isrc -I../src -I../include -I../wayland-egl -Iwayland-eglstream -I/nix/store/scp3wbv44ch6mdrjjhf1d3x0qnb41il6-eglexternalplatform-1.1/include -I/nix/store/h0b47l5z1fqxcgxnh87j4dj72l71zn8h-wayland-1.17.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=gnu99 -Wall -Werror -fvisibility=hidden -DWL_HIDE_DEPRECATED -Wno-pedantic -fPIC -pthread -MD -MQ 'src/25a6634@@nvidia-egl-wayland@sha/wayland-thread.c.o' -MF 'src/25a6634@@nvidia-egl-wayland@sha/wayland-thread.c.o.d' -o 'src/25a6634@@nvidia-egl-wayland@sha/wayland-thread.c.o' -c ../src/wayland-thread.c
In file included from ../src/wayland-thread.c:29:
../include/wayland-egldisplay.h:26:10: fatal error: EGL/egl.h: No such file or directory
 #include <EGL/egl.h>
          ^~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
builder for '/nix/store/iq61rj6j753wx0r9v80ffp6b42rvmpy8-egl-wayland-1.1.4.drv' failed with exit code 1
error: build of '/nix/store/iq61rj6j753wx0r9v80ffp6b42rvmpy8-egl-wayland-1.1.4.drv' failed
```

Are you missing those Khronos EGL headers?

Also, @andersk, I'm using Sway window manager and the firefox-wayland package (on NixOS) and `about:support` says I'm rendering through Wayland.

**Comment by hedning** [comment]
Probably hitting this, mesa removed a bunch of headers in 19.2.2: https://github.com/NixOS/nixpkgs/pull/72999

Looks like `mesa_glu` is correct now, and we need this fix too: https://github.com/NVIDIA/egl-wayland/pull/24

Will make PR.

**Comment by B4dM4n** [comment]
PR is already created: #74325

**Pull request merged**